app/scripts/metamask-controller.js

```diff
@@ -738,10 +774,76 @@ export default class MetamaskController extends EventEmitter {
       );
     });
 
+    this.setupControllerEventSubscriptions();
+
     // TODO:LegacyProvider: Delete
     this.publicConfigStore = this.createPublicConfigStore();
   }
 
+  /**
+   * Sets up BaseController V2 event subscriptions. Currently, this includes
+   * the subscriptions necessary to notify permission subjects of account
+   * changes.
+   *
+   * Some of the subscriptions in this method are ControllerMessenger selector
+   * event subscriptions. See the relevant @metamask/controllers documentation
+   * for more information.
+   *
+   * Note that account-related notifications emitted when the extension
+   * becomes unlocked are handled in MetaMaskController._onUnlock.
+   */
+  setupControllerEventSubscriptions() {
+    const getCurrentAccounts = async (origin, newAccounts) => {
+      return newAccounts.length < 2
+        ? newAccounts
```

Comment:
Wait a minute... why doesn't `getPermittedAccountsByOrigin` return accounts that are sorted correctly, if this is the explanation?

Reply:
I added comments that should clarify the situation here: ef8895171

Reply:
That does clarify everything but that last question; why doesn't `getPermittedAccountsByOrigin` sort the accounts by last selected?

Reply:
They're not sorted because that function is just pulling the accounts directly from the `PermissionController` state, where they are unsorted. Are you asking me to add the sorting logic from the `eth_accounts` implementation to `getPermittedAccountsByOrigin`?

Reply:
I was just wondering if it was considered. I see now that `getPermittedAccountsByOrigin` is a selector, so it wouldn't make a lot of sense for that to be async. This approach seems fine.